### PR TITLE
Mixed cached shadow maps (Cached static + dynamic rendered every frame)

### PR DESCRIPTION
**Pull request by FrancescoC-unity**
**IMPORTANT:** Requires this C++ PR https://ono.unity3d.com/unity/unity/pull-request/110280/_/graphics/concept-of-static-shadow-caster-for-SRP . Whenever that lands I will bump minimum required version of HDRP.

This PR introduces two features. 

The biggest feature is 

**Mixed cached shadow maps**

It is now possible to tag objects as Static Shadow Casters (with the C++ PR above). Selecting the new option Always draw dynamic on the shadows, it is possible to have all static objects cached and dynamic objects rendered on top every frame. 

See the example in the gif, all pillars are marked as static (when moved the shadow is not updated) and the dancing dude is not. You can see how its shadow is updated every frame while the rest of the shadows are not. 

MixedCached

This will allow the cached shadows to be used in many more use cases. Lots of games have static environments but few movable characters and objects, so before those couldn't use cached shadows, but now can for significant perf benefits (both CPU and GPU as most of the objects will not be rendered anymore in shadows). 

Shortly, this work rendering first all static in cached atlas when an update is required, then blit from cache to dynamic atlas and subsequently every frame the dynamic objects are rendered on top of the blitted cached shadows. 

Second feature is: 

**Update upon light movement**
 
There is now a new option to shortcut a usual use case for OnDemand. This will now trigger an update every time a light moves above  a certain threshold set via script (by default really small).

This helps for use cases when a light almost never moves, but you want to update when it does move. 

UpdateLightMovement

**Note:** Rendergraph has not been updated, need a bit of a chat with @JulienIgnace-Unity  to make sure of a change I need to do. 

**What have I tested**: I tested what you can see in the gif, all the light types for both new options. I also tried when the dynamic atlas is full and the mixed cached shadows still work. 
Still I'd love thorough testing on this since I changed a few fundamental things about what determines shadow rendering options.

### CODE REVIEW

File: com.unity.render-pipelines.high-definition/Documentation~/Shadows-in-HDRP.md [hunk 1]
```diff
@@ -108,6 +108,10 @@ When a Light that caches its shadows renders its shadow map for the first time,
...
 For a Light that does cache its shadows, if you disable it or set its **Update Mode** to **Every Frame**, you can tell HDRP to preserve the Light's shadow map's place in the cached shadow atlas. This means that, if you enable the Light again, HDRP does not need to re-render the shadow map or place it into a shadow atlas. For information on how to make a Light preserve its shadow map's place in the cached shadow atlas, see [Preserving shadow atlas placement](#preserving-shadow-atlas-placement).
 
+As a shortcut for a common case, HDRP offers an option to automatically trigger an update when either the position or rotation of a light changes above a certain threshold. This option can be selected in the Shadow settings section on the Light as **Update on light movement**.
```
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/Documentation~/Shadows-in-HDRP.md:111]: As a shortcut for a common case, HDRP offers an option to automatically trigger an update when either the position or rotation of a light changes above a certain threshold. To enable this option, select a [Light](Light-Component.md) and, in the **Shadow** section of its Inspector, enable **Update on light movement**.

File: com.unity.render-pipelines.high-definition/Documentation~/Shadows-in-HDRP.md [hunk 2]
```diff
@@ -108,6 +108,10 @@ When a Light that caches its shadows renders its shadow map for the first time,
...
 For a Light that does cache its shadows, if you disable it or set its **Update Mode** to **Every Frame**, you can tell HDRP to preserve the Light's shadow map's place in the cached shadow atlas. This means that, if you enable the Light again, HDRP does not need to re-render the shadow map or place it into a shadow atlas. For information on how to make a Light preserve its shadow map's place in the cached shadow atlas, see [Preserving shadow atlas placement](#preserving-shadow-atlas-placement).
 
+As a shortcut for a common case, HDRP offers an option to automatically trigger an update when either the position or rotation of a light changes above a certain threshold. This option can be selected in the Shadow settings section on the Light as **Update on light movement**. 
+The threshold that are used to determine how much a light needs to move or rotate to trigger an update can be customized per light with the properties
```
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/Documentation~/Shadows-in-HDRP.md:112]: You can customize the threshold that HDRP uses to determine how much a light needs to move or rotate to trigger an update. To do this, use the:

File: com.unity.render-pipelines.high-definition/Documentation~/Shadows-in-HDRP.md [hunk 3]
```diff
@@ -108,6 +108,10 @@ When a Light that caches its shadows renders its shadow map for the first time,
...
 For a Light that does cache its shadows, if you disable it or set its **Update Mode** to **Every Frame**, you can tell HDRP to preserve the Light's shadow map's place in the cached shadow atlas. This means that, if you enable the Light again, HDRP does not need to re-render the shadow map or place it into a shadow atlas. For information on how to make a Light preserve its shadow map's place in the cached shadow atlas, see [Preserving shadow atlas placement](#preserving-shadow-atlas-placement).
 
+As a shortcut for a common case, HDRP offers an option to automatically trigger an update when either the position or rotation of a light changes above a certain threshold. This option can be selected in the Shadow settings section on the Light as **Update on light movement**. 
+The threshold that are used to determine how much a light needs to move or rotate to trigger an update can be customized per light with the properties 
+`cachedShadowTranslationUpdateThreshold` and `cachedShadowAngleUpdateThreshold` on the **HDAdditionalLightData** component. Note that point lights ignore the angle differences when determining if an update is needed with this mode.
```
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/Documentation~/Shadows-in-HDRP.md:113]: `cachedShadowTranslationUpdateThreshold` and `cachedShadowAngleUpdateThreshold` properties on the Light's **HDAdditionalLightData** component. Note that point lights ignore the angle differences when determining if they need to perform an update in this mode.

File: com.unity.render-pipelines.high-definition/Documentation~/Shadows-in-HDRP.md [hunk 4]
```diff
@@ -131,6 +135,16 @@ Note that this causes HDRP to mark all the shadow maps in the atlas as dirty whi
...
 If you plan to only temporarily set a Light's **Update Mode** to **Every Frame** and want to set it back to **On Enable** or **On Demand** later, you can preserve the Light's shadow map placement in its atlas. This is useful, for example, if you want HDRP to cache a far away Light's shadow map, but update it every frame when it gets close to the [Camera](HDRP-Camera.md). To do this, access the Light's **HDAdditionalLightData** component and enable the **preserveCachedShadow** property. If this property is set to `true`, HDRP preserves the Light's shadow map's space in its shadow atlas. Note that even if this property is enabled, if you destroy the Light, it loses its placement in the shadow atlas.
 
+### Mixed Cached Shadow Maps
+
+It is possible to cache only a portion of the shadow map. In order to do so, you must check the **Always draw dynamic** option in the shadow settings of the light and then tick the **Static Shadow Caster** checkbox all renderers that will be cached in shadows (in the Renderer component).
```
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/Documentation~/Shadows-in-HDRP.md:140]: It is possible to cache only a portion of the shadow map. To do this, enable the **Always draw dynamic** option in the [Light's](Light-Component.md) shadow settings and then enable the **Static Shadow Caster** option for all Renderers to cache shadows for.

File: com.unity.render-pipelines.high-definition/Documentation~/Shadows-in-HDRP.md [hunk 5]
```diff
@@ -131,6 +135,16 @@ Note that this causes HDRP to mark all the shadow maps in the atlas as dirty whi
...
 
+### Mixed Cached Shadow Maps
+
+It is possible to cache only a portion of the shadow map. In order to do so, you must check the **Always draw dynamic** option in the shadow settings of the light and then tick the **Static Shadow Caster** checkbox all renderers that will be cached in shadows (in the Renderer component). 
+With such setup, all static shadow casters are rendered in the shadow map only whenever an explicit update is requested (or only on light enable in case of **OnEnable** update mode), while the other casters (dynamic shadow casters) are drawn each frame.
```
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/Documentation~/Shadows-in-HDRP.md:141]: With this setup, HDRP only renders static shadow casters into the shadow map when an explicit update is requested (or, if the Light uses the **OnEnable** update mode, only when the Light enables) whereas, for dynamic shadow casters, HDRP renders them into their respective shadow maps each frame.

File: com.unity.render-pipelines.high-definition/Documentation~/Shadows-in-HDRP.md [hunk 6]
```diff
@@ -131,6 +135,16 @@ Note that this causes HDRP to mark all the shadow maps in the atlas as dirty whi
...
 
+### Mixed Cached Shadow Maps
+
+It is possible to cache only a portion of the shadow map. In order to do so, you must check the **Always draw dynamic** option in the shadow settings of the light and then tick the **Static Shadow Caster** checkbox all renderers that will be cached in shadows (in the Renderer component). 
+With such setup, all static shadow casters are rendered in the shadow map only whenever an explicit update is requested (or only on light enable in case of **OnEnable** update mode), while the other casters (dynamic shadow casters) are drawn each frame. 
+This setup is particularly useful if for example your environment is mostly static and the light doesn't move, but there are few dynamic objects that will need shadows to be cast by such lights. In such scenarios, setting the light as having a mixed cached shadow map will greatly improve performance both on CPU and GPU.
```
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/Documentation~/Shadows-in-HDRP.md:142]: This setup is particularly useful if your environment consists of mostly static GameObjects and the lights do not move, but there are few dynamic GameObjects that you want the static lights to cast shadows for. In such scenarios, setting the light to have a mixed cached shadow map greatly improves performance both on CPU and GPU.

File: com.unity.render-pipelines.high-definition/Documentation~/Shadows-in-HDRP.md [hunk 7]
```diff
@@ -131,6 +135,16 @@ Note that this causes HDRP to mark all the shadow maps in the atlas as dirty whi
...
+With such setup, all static shadow casters are rendered in the shadow map only whenever an explicit update is requested (or only on light enable in case of **OnEnable** update mode), while the other casters (dynamic shadow casters) are drawn each frame. 
+This setup is particularly useful if for example your environment is mostly static and the light doesn't move, but there are few dynamic objects that will need shadows to be cast by such lights. In such scenarios, setting the light as having a mixed cached shadow map will greatly improve performance both on CPU and GPU. 
+
+Please note that as part of implementation details, if a shadow is set up as mixed cached, every frame a blit from the cached shadow map to the dynamic atlas is performed. This is important to keep in mind both for the extra runtime cost of the blit in itself, but also to understand that in terms of memory a single shadow map will require space in both atlases.
```
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/Documentation~/Shadows-in-HDRP.md:144]: Note that, due to implementation details, if you set up a shadow to be mixed cached, HDRP performs a blit from the cached shadow map to the dynamic atlas. This is important to keep in mind both for the extra runtime cost of the blit in itself, but also to understand that, in terms of memory, a single shadow map requires space in both atlases.

File: com.unity.render-pipelines.high-definition/Documentation~/Shadows-in-HDRP.md [hunk 8]
```diff
@@ -131,6 +135,16 @@ Note that this causes HDRP to mark all the shadow maps in the atlas as dirty whi
...
+This setup is particularly useful if for example your environment is mostly static and the light doesn't move, but there are few dynamic objects that will need shadows to be cast by such lights. In such scenarios, setting the light as having a mixed cached shadow map will greatly improve performance both on CPU and GPU. 
+
+Please note that as part of implementation details, if a shadow is set up as mixed cached, every frame a blit from the cached shadow map to the dynamic atlas is performed. This is important to keep in mind both for the extra runtime cost of the blit in itself, but also to understand that in terms of memory a single shadow map will require space in both atlases. 
+
+Another important note, is due to implementation details, if a light with mixed cached shadows moves and the cached counterpart is not updated, the result will look wrong. In such cases either select the **Update on light movement** or the update mode to **OnDemand** and make sure an update is triggered when desired.
```
Comment (JordanL8) [inline com.unity.render-pipelines.high-definition/Documentation~/Shadows-in-HDRP.md:146]: Another important note, also due to implementation details, if a Light with mixed cached shadows moves and you do not update the cached counterpart, the result looks wrong. In such cases either enable the Light's **Update on light movement** option or set the Light's update mode to **OnDemand** and make sure to trigger an update when you move the Light.

**Comment by FrancescoC-unity** [comment]
Fixed up docs as suggested @JordanL8

**Comment by iM0ve** [comment]
**Review in Progress**


**ISSUES**
1) Tooltip has a space before the text

image

2) Setup is a bit confusing. Do we have a restriction that's why it's like that?

From the documentation, it notes that this feature is useful because most of the time MAJORITY of the assets are static but you have a few that you want dynamic shadows for. But the implementation for the user is completely different, by checking **Always draw dynamic** on a light - all of your shadows become realtime, forcing you to go over every object and fix them.

Can the **Always draw dynamic** checbox be removed or we need it?
If by default **Always draw dynamic* is always on(remove parameter) and the **Static shadow caster** is also On by default (dont remove parameter). Then usage would be very simplified, all you have to do is uncheck **Static shadow caster** for objects you want realtime shadows for and thats it. Again Im just looking at this from usage convenience, maybe there is a technical reason we can not implement it like this.

**Test coverage**
* Tooltips
* Documentation

**Comment by FrancescoC-unity** [comment]
1- Will fix, though it is on C++ side so it will be a bit of a mess :-) Will see for next release. 

2- Several points to make here: 

 The checkbox we absolutely need otherwise all cached shadows will be mixed and it is not what we want. We need to have a discrimination of the two types. Sometimes you are fine with dynamic stuff being rendered for one light and not for another and to differentiate you need a checkbox on the light. 

  The checkbox on the renderer is engine side and it is going to be used by all SRPs, we cannot make it default (changing default behaviour) for HDRP usages. So while I agree, if it was hdrp only I would've made that on by default, we cannot really do that. 

 3- The static checkbox up there is lost at runtime, so it is unusable sadly. Again, the reason for the naming is due to it being a generic engine side thing and its implementation is generic to any Shadow map draw call that is done via SRP, so cannot really  make it reference cached shadow maps. The tooltip is there to show a potential usage linking to what HDRP does.

**Comment by iM0ve** [comment]
Thanks, added 4 and 5.

**Comment by FrancescoC-unity** [comment]
4- I agree that it is a bit bad, but unfortunately we can't fix this case as we don't have a list of game objects that are in each state on C# nor easily in C++. If we ever will have scriptable culling then this could change at some point; as of now sadly not. 

5- I think i know what's the issue! I am in vacation, but will check and fix when back.

**Review by JulienIgnace-Unity — APPROVED** [review]

**Comment by FrancescoC-unity** [comment]
I disabled the directional light case, to make it work we need even more memory. Waiting for the RG to be main case to do that. 
It is the least useful case anyway (since directional light shadows need frequent re-updates). 

Ping @iM0ve re the bug above on the directional light.

**Review by iM0ve — CHANGES_REQUESTED** [review]
The issue #5 still reproduces for me. 
I have:
* Pulled the disable/fix 
* Deleted the Library folder
* Tried turning Directional light on/off

asdgewg

**Review by iM0ve — APPROVED** [review]
All issues fixed or addressed

**Review by sebastienlagarde — APPROVED** [review]

**Pull request merged**